internal/node/link_node_modules.js

```diff
@@ -0,0 +1,65 @@
+/**
+ * @fileoverview Creates a node_modules directory in the current working directory
+ * and symlinks in the node modules needed to run a program.
+ * This replaces the need for custom module resolution logic inside the process.
+ */
+const fs = require('fs');
+const path = require('path');
+
+function debug(...m) {
+  if (!!process.env['DEBUG']) console.error('[link_node_modules.js]', ...m);
+}
+
+function ls_r(d) {
+  for (const p of fs.readdirSync(d)) {
+    const f = path.join(d, p);
+    debug('  ', f);
+    if (fs.statSync(f).isDirectory()) {
+      ls_r(f);
+    }
+  }
+}
+
+function symlink(target, path) {
+  debug(`fs.symlinkSync(${target}, ${path})`);
+  fs.symlinkSync(target, path);
+}
+
+function main(args) {
+  const [modulesManifest] = args;
+  const {root, modules} = JSON.parse(fs.readFileSync(modulesManifest));
+  debug('read module manifest, node_modules root is', root, 'with first-party packages', modules);
+  if (!root) {
+    // FIXME: this case with only first-party modules should work too!!
+    debug('Since there is no third-party node_modules tree, we do nothing');
+    return;
+  }
+
+  // First, symlink from a local `$PWD/node_modules directory to the root of the third-party
+  // installed node_modules
+  if (!fs.existsSync('node_modules')) {
+    symlink(root, 'node_modules');
+  }
+
+  const symlinkRelativeTarget = path.relative(root, '.');
+  process.chdir(root);
+
+  // Now add symlinks to each of our first-party packages so they appear under the node_modules tree
+  for (const m of Object.keys(modules)) {
+    if (fs.existsSync(m)) continue;
```

Comment:
same with this `existsSync` convert it to the async version and do this loop in parallel